ui/src/treegrid.tsx

```diff
@@ -0,0 +1,247 @@
+  const [state, updateState] = useImmer<TreeGridState>(
+    new Map<TreeGridId, TreeGridItemState>()
+  );
+
+  const toggleExpanded = useCallback((id: TreeGridId) => {
+    updateState((draft) => {
+      const itemState = draft.get(id) || {
+        status: Status.Collapsed,
+      };
+      switch (itemState.status) {
+        case Status.Loading:
+          return;
+        case Status.Expanded:
+          itemState.status = Status.Collapsed;
+          break;
+        default:
+          if (!props.loadChildren || !!itemState.loaded) {
+            itemState.status = Status.Expanded;
+          } else {
+            itemState.status = Status.Loading;
+            props
+              .loadChildren(id)
+              .then(() => {
+                updateState((draft) => {
+                  draft.set(id, {
+                    status: Status.Expanded,
+                    loaded: true,
+                  });
+                });
+              })
+              .catch((error) => {
+                updateState((draft) => {
+                  draft.set(id, {
+                    status: Status.Failed,
+                    errorMessage: error,
+                  });
+                });
+              });
+          }
+      }
+      draft.set(id, itemState);
+    });
+  }, []);
+
+  return (
+    <>
+      <table
+        style={{
+          tableLayout: "fixed",
+        }}
+      >
+        <thead>
+          <tr>
+            {props.columns.map((col, i) => (
+              <td
+                key={i}
+                style={{
+                  ...(col.width && {
+                    maxWidth: col.width,
+                    width: col.width,
+                    minWidth: col.width,
+                  }),
+                }}
+              >
+                <Typography variant="h6">{col.title}</Typography>
+              </td>
+            ))}
+          </tr>
+        </thead>
+      </table>
+      <div style={{ overflowY: "auto", display: "block" }}>
+        <table
+          style={{
+            tableLayout: "fixed",
+          }}
+        >
+          <tbody>
+            {renderChildren(props, state, toggleExpanded, "root", 0, false)}
+          </tbody>
+        </table>
+      </div>
+    </>
+  );
+}
+
+function renderChildren(
+  props: TreeGridProps,
+  state: TreeGridState,
+  toggleExpanded: (id: TreeGridId) => void,
+  id: TreeGridId,
+  indent: number,
+  collapse: boolean
+): JSX.Element[] {
+  const results: JSX.Element[] = [];
+
+  props.data[id]?.children?.forEach((childId) => {
+    const child = props.data[childId];
+    if (!child) {
+      return;
+    }
+    const childState = state.get(childId);
+
+    results.push(
+      <tr
+        key={childId}
+        style={{
+          ...(collapse && { visibility: "collapse" }),
+        }}
+      >
+        {props.columns.map((col, i) => {
+          let value = child.data[col.key] || "";
+          let title = "";
+          if (!(value instanceof Object)) {
```

Comment:
Can you add a comment here? Is this attempting to check for JSX elements?

Reply:
Yeah, done.